grpc/server/src/main/java/io/helidon/grpc/server/ProtoReflectionService.java

```diff
@@ -0,0 +1,579 @@
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+import com.google.protobuf.Any;
+import com.google.protobuf.DescriptorProtos;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import com.google.protobuf.Descriptors.MethodDescriptor;
+import com.google.protobuf.Descriptors.ServiceDescriptor;
+import io.grpc.BindableService;
+import io.grpc.InternalServer;
+import io.grpc.Server;
+import io.grpc.ServerMethodDefinition;
+import io.grpc.ServerServiceDefinition;
+import io.grpc.Status;
+import io.grpc.protobuf.ProtoFileDescriptorSupplier;
+import io.grpc.reflection.v1alpha.ErrorResponse;
+import io.grpc.reflection.v1alpha.ExtensionNumberResponse;
+import io.grpc.reflection.v1alpha.ExtensionRequest;
+import io.grpc.reflection.v1alpha.FileDescriptorResponse;
+import io.grpc.reflection.v1alpha.ListServiceResponse;
+import io.grpc.reflection.v1alpha.ServerReflectionGrpc;
+import io.grpc.reflection.v1alpha.ServerReflectionRequest;
+import io.grpc.reflection.v1alpha.ServerReflectionResponse;
+import io.grpc.reflection.v1alpha.ServiceResponse;
+import io.grpc.stub.ServerCallStreamObserver;
+import io.grpc.stub.StreamObserver;
+
+import static io.grpc.MethodDescriptor.MethodType.BIDI_STREAMING;
+import static io.grpc.MethodDescriptor.MethodType.CLIENT_STREAMING;
+import static io.grpc.MethodDescriptor.MethodType.SERVER_STREAMING;
+
+/**
+ * Provides a reflection service for Protobuf services (including the reflection service itself).
+ *
+ * <p>Separately tracks mutable and immutable services. Throws an exception if either group of
+ * services contains multiple Protobuf files with declarations of the same service, method, type, or
+ * extension.
+ */
+public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase {
+
+    private final Object lock = new Object();
+
+    private final Map<Server, ServerReflectionIndex> serverReflectionIndexes = new WeakHashMap<>();
+
+    private ProtoReflectionService() {
+    }
+
+    /**
+     * Creates a instance of {@link ProtoReflectionService}.
+     * @return a service instance
+     */
+    public static BindableService newInstance() {
+        return new ProtoReflectionService();
+    }
+
+    /**
+     * Retrieves the index for services of the server that dispatches the current call. Computes
+     * one if not exist. The index is updated if any changes to the server's mutable services are
+     * detected. A change is any addition or removal in the set of file descriptors attached to the
+     * mutable services or a change in the service names.
+     */
+    private ServerReflectionIndex getRefreshedIndex() {
```

Comment:
How often is this called? Hard synchronization is expensive... 
Do we need to do it as pull approach (e.g. do we need to query the server for all services, as opposed to getting events about mutable services being changed)?

Reply:
In reality, only once, so it shouldn't be a big deal.

Services don't really change at runtime, so there shouldn't be any changes that need refreshing.